fdbserver/ConfigBroadcaster.actor.cpp

```diff
@@ -49,101 +49,56 @@ void remove(Container& container, K const& k) {
-					for (const auto& [token, req] : tokenToRequest) {
-						if (req.lastSeenVersion < change.version) {
-							tokenSet.insert(token);
-						}
-					}
-				} else {
-					Key configClass = change.mutation.getConfigClass().get();
-					if (configClassToTokens.count(configClass)) {
-						auto tokens = get(configClassToTokens, Key(change.mutation.getConfigClass().get()));
-						for (const auto& token : tokens) {
-							auto req = get(tokenToRequest, token);
-							if (req.lastSeenVersion < change.version) {
-								tokenSet.insert(token);
-							} else {
-								TEST(true); // Worker is ahead of config broadcaster
-							}
-						}
-					}
-				}
-			}
-			std::vector<Req> result;
-			for (const auto& token : tokenSet) {
-				result.push_back(get(tokenToRequest, token));
-			}
-			return result;
+	// Holds information about each client connected to the broadcaster.
+	struct BroadcastClientDetails {
+		// Triggered when the worker dies.
+		Future<Void> watcher;
+		ConfigClassSet configClassSet;
+		Version lastSeenVersion;
+		ConfigBroadcastInterface broadcastInterface;
+
+		bool operator==(BroadcastClientDetails const& rhs) const {
+			return configClassSet == rhs.configClassSet && lastSeenVersion == rhs.lastSeenVersion &&
+			       broadcastInterface == rhs.broadcastInterface;
 		}
+		bool operator!=(BroadcastClientDetails const& rhs) const { return !(*this == rhs); }
+	};
 
-		std::vector<Req> getOutdatedRequests(Version newSnapshotVersion) {
-			std::vector<Req> result;
-			for (const auto& [token, req] : tokenToRequest) {
-				if (req.lastSeenVersion < newSnapshotVersion) {
-					result.push_back(req);
-				}
-			}
-			return result;
-		}
-
-		void removeRequest(Req const& req) {
-			auto token = req.reply.getEndpoint().token;
-			for (const auto& configClass : req.configClassSet.getClasses()) {
-				remove(get(configClassToTokens, configClass), token);
-				// TODO: Don't leak config classes
-			}
-			remove(tokenToRequest, token);
-		}
-	} pending;
 	std::map<ConfigKey, KnobValue> snapshot;
 	std::deque<VersionedConfigMutation> mutationHistory;
 	std::deque<VersionedConfigCommitAnnotation> annotationHistory;
 	Version lastCompactedVersion;
 	Version mostRecentVersion;
 	std::unique_ptr<IConfigConsumer> consumer;
 	ActorCollection actors{ false };
+	std::vector<BroadcastClientDetails> clients;
 
 	UID id;
 	CounterCollection cc;
 	Counter compactRequest;
-	mutable Counter successfulChangeRequest;
+	Counter successfulChangeRequest;
 	Counter failedChangeRequest;
 	Counter snapshotRequest;
 	Future<Void> logger;
 
 	template <class Changes>
-	void sendChangesReply(ConfigBroadcastFollowerGetChangesRequest const& req, Changes const& changes) const {
-		ASSERT_LT(req.lastSeenVersion, mostRecentVersion);
-		ConfigBroadcastFollowerGetChangesReply reply;
-		reply.mostRecentVersion = mostRecentVersion;
+	Future<Void> pushChanges(BroadcastClientDetails& client, Changes const& changes) {
+		if (client.watcher.isReady()) {
```

Comment:
Is this check necessary? Shouldn't `waitForFailure` callback have run already if `client.watcher` is ready?

Reply:
This is probably not necessary, I added this before `waitForFailure`. However, I have observed that `waitForFailure` can take a few seconds to trigger after killing a worker, in which case this check would still be necessary. I will add a unit test to kill a worker then immediately set a knob and see if this check is necessary.

Reply:
It may take a few seconds for the information that the worker is dead to reach the `client.watcher` future, but during this time, the above `client.watcher.isReady()` check will also be false.